### ISSUE TITLE
Not sure about final embeddings vectors 

### ISSUE THREAD

**Issue by MaxenceQueyrel** [bentrevett/pytorch-seq2seq#63]
Hello,

Thank you very much for you tutorials.
I think I have an issue in my use case. 
I use the seq2seq with attention model and I want to get embeddings of a sequence. I don't make translation like french -> english, but I want to reconstruct the initial sentences like french -> french. I add some noise in the input sequence to make it more difficult to recover.
When the model is training, the loss is decreasing but the embeddings vectors at the end look like [1., -1., -1, 0.999, 1, -1...] with almost all values equal to -1 and 1.
Can you please help me to interpret this ? Is it a normal case or is it a form of overfitting ?

**Comment by bentrevett** [comment]
Which model are you using? By "embeddings" do you mean the output of the `nn.Embedding` layer? Have you tried outputting the reconstructed sequences and seeing how close they are to the original sentences?

The values output by any layer should be around zero and larger values usually do usually imply overfitting, but having them all be close to either +/- 1 does seem a bit odd.

**Comment by MaxenceQueyrel** [comment]
Sorry for the late answer, I wanted to run a new model and check what you asked for.

I am using the algorithm from Neural Machine Translation by Jointly Learning to Align and Translate tutorial.
By embeddings I mean the embeddings of the sentence, so the one outputed by the encoder layer.

When I reconstruct a sentence that I have created (so not in the training set) it seems to work with few mistakes but when I check the embeddings they are close to -1 and 1, as you said this is weird.

So if you think it is overfitting, and I think so too, what do you recommend to me to avoid this phenomenon ? Or which experiments to test do you recommend to me ?

Thank you.

**Comment by bentrevett** [comment]
@MaxenceQueyrel I have trained an English-to-English reconstructor: https://gist.github.com/bentrevett/6986196132f253cb48de88e90605901f

As you can see, after training for 5 epochs I use `get_embedding_values` to get the `outputs` from a single batch of sentences and then plot them on a histogram. My values look reasonable, i.e. a Gaussian distribution, but they do have small spikes at +1 and -1. 

The only thing I haven't added is the noise which you said you were adding, perhaps that could be a reason why you have these spikes? 

Also, can you try and create a histogram with what you're getting and post it?

**Comment by MaxenceQueyrel** [comment]
@bentrevett Thank you a lot for this notebook. At the beginning it looks I did the same as you to collect the embeddings. But after I saw my mistake, I have inverted the name of two variables, so I was printing the hidden variable and not the outputs one. Because of the tanh function, the values were between -1 and 1, I should have seen this error earlier sorry !

For the noise I put a probability of 0.05 to permute two words, and a probability of 0.02 to remove one word. When I test both methods and check similarities it seems to be equivalent. Maybe it can be interesting to try both approaches for a specific NLP task and compare the results.

Thank you again to helping me with this issue !
